File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/Diff.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface DiffLib: Library {
    fun Diff_destroy(handle: Pointer)
    fun Diff_compute(a: Pointer, b: Pointer, preset: Slice): ResultPointerInt
    fun Diff_compute_with_opts(a: Pointer, b: Pointer, preset: Slice, costAdd: Int, costDelete: Int, costChange: Int, costSwap: Int, symmetry: Slice): ResultPointerInt
    fun Diff_from_json(json: Slice): ResultPointerInt
    fun Diff_distance(handle: Pointer): FFIUint64
    fun Diff_support(handle: Pointer): Float
    fun Diff_to_json(handle: Pointer, write: Pointer): Unit
    fun Diff_summary_json(handle: Pointer, write: Pointer): Unit
    fun Diff_added(handle: Pointer): Pointer
    fun Diff_removed(handle: Pointer): Pointer
    fun Diff_changed(handle: Pointer): Pointer
    fun Diff_swapped(handle: Pointer): Pointer
    fun Diff_markers(handle: Pointer): Pointer
    fun Diff_to_overlay_glb_b64(handle: Pointer, afterGlb: Slice, write: Pointer): ResultUnitInt
}
/** A computed diff between two schematics.
*/
class Diff internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class DiffCleaner(val handle: Pointer, val lib: DiffLib) : Runnable {
        override fun run() {
            lib.Diff_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, Diff.DiffCleaner(handle, Diff.lib));
    }

    companion object {
        internal val libClass: Class<DiffLib> = DiffLib::class.java
        internal val lib: DiffLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        /** Diff two schematics with the given preset (default cost model).
        *Errors with `InvalidArgument` on an unknown preset.
        */
        fun compute(a: Schematic, b: Schematic, preset: String): Result<Diff> {
            val presetSliceMemory = PrimitiveArrayTools.borrowUtf8(preset)
            
            val returnVal = lib.Diff_compute(a.handle, b.handle, presetSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = Diff(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                presetSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** Diff two schematics with optional cost/symmetry overrides. Negative
        *cost ints mean "unset" (use the preset default); an empty `symmetry`
        *string means "unset". Errors with `InvalidArgument` on an unknown
        *preset or symmetry name.
        */
        fun computeWithOpts(a: Schematic, b: Schematic, preset: String, costAdd: Int, costDelete: Int, costChange: Int, costSwap: Int, symmetry: String): Result<Diff> {
            val presetSliceMemory = PrimitiveArrayTools.borrowUtf8(preset)
            val symmetrySliceMemory = PrimitiveArrayTools.borrowUtf8(symmetry)
            
            val returnVal = lib.Diff_compute_with_opts(a.handle, b.handle, presetSliceMemory.slice, costAdd, costDelete, costChange, costSwap, symmetrySliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = Diff(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                presetSliceMemory.close()
                symmetrySliceMemory.close()
            }
        }
        @JvmStatic
        
        /** Reconstruct a diff from its JSON representation.
        */
        fun fromJson(json: String): Result<Diff> {
            val jsonSliceMemory = PrimitiveArrayTools.borrowUtf8(json)
            
            val returnVal = lib.Diff_from_json(jsonSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = Diff(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                jsonSliceMemory.close()
            }
        }
    }
    
    /** The edit distance of the diff.
    */
    fun distance(): ULong {
        
        val returnVal = lib.Diff_distance(handle);
        return (returnVal.toULong())
    }
    
    /** The support (alignment confidence) of the diff.
    */
    fun support(): Float {
        
        val returnVal = lib.Diff_support(handle);
        return (returnVal)
    }
    
    /** Serialize the diff to its full JSON representation.
    */
    fun toJson(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Diff_to_json(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    /** Serialize the diff to its compact summary JSON.
    */
    fun summaryJson(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Diff_summary_json(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    /** A new schematic containing only the blocks added in this diff.
    */
    fun added(): Schematic {
        
        val returnVal = lib.Diff_added(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Schematic(handle, selfEdges, true)
        return returnOpaque
    }
    
    /** A new schematic containing only the blocks removed in this diff.
    */
    fun removed(): Schematic {
        
        val returnVal = lib.Diff_removed(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Schematic(handle, selfEdges, true)
        return returnOpaque
    }
    
    /** A new schematic containing only the blocks changed in this diff.
    */
    fun changed(): Schematic {
        
        val returnVal = lib.Diff_changed(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Schematic(handle, selfEdges, true)
        return returnOpaque
    }
    
    /** A new schematic containing only the blocks swapped in this diff.
    */
    fun swapped(): Schematic {
        
        val returnVal = lib.Diff_swapped(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Schematic(handle, selfEdges, true)
        return returnOpaque
    }
    
    /** A new schematic with marker blocks summarizing this diff.
    */
    fun markers(): Schematic {
        
        val returnVal = lib.Diff_markers(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Schematic(handle, selfEdges, true)
        return returnOpaque
    }
    
    /** Render a diff overlay on top of an "after" GLB buffer, returning the
    *new GLB as base64 (PORTING rule 6). Requires the `meshing` feature;
    *errors with `Mesh` when compiled without it.
    */
    fun toOverlayGlbB64(afterGlb: UByteArray): Result<String> {
        val afterGlbSliceMemory = PrimitiveArrayTools.borrow(afterGlb)
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Diff_to_overlay_glb_b64(handle, afterGlbSliceMemory.slice, write);
        try {
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                
                val returnString = DW.writeToString(write)
                return returnString.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        } finally {
            afterGlbSliceMemory.close()
        }
    }

}

File: bindings/php/Diff.php
<?php
namespace Stencil;

final class Diff {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned) {
        $this->ptr = $ptr;
        $this->owned = $owned;
    }

    public static function compute( $a,  $b, string $preset) {
        $__n2 = strlen($preset);
        $__view2 = Lib::ffi()->new('DiplomatStringView');
        if ($__n2 > 0) {
            $__buf2 = Lib::ffi()->new("uint8_t[" . $__n2 . "]", false);
            \FFI::memcpy($__buf2, $preset, $__n2);
            $__view2->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf2[0]));
        } else {
            $__view2->data = null;
        }
        $__view2->len = $__n2;
        $result = Lib::ffi()->Diff_compute($a->ptr, $b->ptr, $__view2);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new Diff($result->ok, true);
    }

    public static function computeWithOpts( $a,  $b, string $preset,  $cost_add,  $cost_delete,  $cost_change,  $cost_swap, string $symmetry) {
        $__n2 = strlen($preset);
        $__view2 = Lib::ffi()->new('DiplomatStringView');
        if ($__n2 > 0) {
            $__buf2 = Lib::ffi()->new("uint8_t[" . $__n2 . "]", false);
            \FFI::memcpy($__buf2, $preset, $__n2);
            $__view2->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf2[0]));
        } else {
            $__view2->data = null;
        }
        $__view2->len = $__n2;
        $__n7 = strlen($symmetry);
        $__view7 = Lib::ffi()->new('DiplomatStringView');
        if ($__n7 > 0) {
            $__buf7 = Lib::ffi()->new("uint8_t[" . $__n7 . "]", false);
            \FFI::memcpy($__buf7, $symmetry, $__n7);
            $__view7->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf7[0]));
        } else {
            $__view7->data = null;
        }
        $__view7->len = $__n7;
        $result = Lib::ffi()->Diff_compute_with_opts($a->ptr, $b->ptr, $__view2, $cost_add, $cost_delete, $cost_change, $cost_swap, $__view7);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new Diff($result->ok, true);
    }

    public static function fromJson(string $json) {
        $__n0 = strlen($json);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $json, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->Diff_from_json($__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new Diff($result->ok, true);
    }

    public function distance() {
        $ret = Lib::ffi()->Diff_distance($this->ptr);
        return $ret;
    }

    public function support() {
        $ret = Lib::ffi()->Diff_support($this->ptr);
        return $ret;
    }

    public function toJson() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        Lib::ffi()->Diff_to_json($this->ptr, $write);
        return Lib::readAndFreeWrite($write);
    }

    public function summaryJson() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        Lib::ffi()->Diff_summary_json($this->ptr, $write);
        return Lib::readAndFreeWrite($write);
    }

    public function added() {
        $ret = Lib::ffi()->Diff_added($this->ptr);
        return new Schematic($ret, true);
    }

    public function removed() {
        $ret = Lib::ffi()->Diff_removed($this->ptr);
        return new Schematic($ret, true);
    }

    public function changed() {
        $ret = Lib::ffi()->Diff_changed($this->ptr);
        return new Schematic($ret, true);
    }

    public function swapped() {
        $ret = Lib::ffi()->Diff_swapped($this->ptr);
        return new Schematic($ret, true);
    }

    public function markers() {
        $ret = Lib::ffi()->Diff_markers($this->ptr);
        return new Schematic($ret, true);
    }

    public function toOverlayGlbB64(array $after_glb) {
        $__n0 = count($after_glb);
        $__view0 = Lib::ffi()->new('DiplomatU8View');
        if ($__n0 > 0) {
            $__arr0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            foreach ($after_glb as $__i0 => $__v0) { $__arr0[$__i0] = $__v0; }
            $__view0->data = \FFI::addr($__arr0[0]);
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Diff_to_overlay_glb_b64($this->ptr, $__view0, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->Diff_destroy($this->ptr);
        }
    }
}
